### PR TITLE
fix `isVisible` false positives from deep nesting or alternate means

### LINKED ISSUE #33914: isVisible utility function returning false positives
The current implementation of [`isVisible`](https://github.com/twbs/bootstrap/blob/main/js/src/util/index.js#L137) is currently not thorough enough. It is claiming certain elements are visible when they clearly are not. [Here is a demo](https://jsfiddle.net/q4s7Lcr1/) of the function thinking that a dropdown item is visible, even when the dropdown is closed.

### PR DESCRIPTION
**Pull request by RyanBerliner**
fixes #33914

### CODE REVIEW

File: js/src/util/index.js [hunk 1]
```diff
@@ -135,20 +135,26 @@ const typeCheckConfig = (componentName, config, configTypes) => {
 }
 
 const isVisible = element => {
-  if (!element) {
+  const eligibleElement = el => el && el.style
+
+  if (!eligibleElement(element)) {
```
Comment (XhmikosR) [inline js/src/util/index.js:140]: Should we combine the two `if`s? Maybe add a line break to make it clearer, though.

File: js/src/util/index.js [hunk 2]
```diff
@@ -135,20 +135,26 @@ const typeCheckConfig = (componentName, config, configTypes) => {
...
+  if (element.tagName !== 'AREA' && element.getClientRects().length === 0) {
     return false
   }
 
-  if (element.style && element.parentNode && element.parentNode.style) {
-    const elementStyle = getComputedStyle(element)
-    const parentNodeStyle = getComputedStyle(element.parentNode)
+  while (eligibleElement(element)) {
+    const { display, visibility } = getComputedStyle(element)
+    if (display === 'none' || visibility === 'hidden') {
```
Comment (alpadev) [inline js/src/util/index.js:150]: In theory this could return false negatives for `visibility` because an element can be visible by using `visibility: visible` while the parent has `visibility: hidden`.

File: js/src/util/index.js (hunk 1 above)
Comment (RyanBerliner) [reply js/src/util/index.js:140]: Do you think this looks more clear?

```javascript
if (
  !eligibleElement(element) ||
  (element.tagName !== 'AREA' && element.getClientRects().length === 0)
) {
  return false
}
```

**Review by alpadev — APPROVED** [review]
Can't judge for the if-condition, but everything else LGTM, thank you! :+1:

**Comment by alpadev** [comment]
@RyanBerliner Hm wouldn't be checking `getClientRects()` enough to see if an element (or an ancestor of it) has `display: none`? 🤔

File: js/src/util/index.js [hunk 3]
```diff
@@ -135,20 +135,33 @@ const typeCheckConfig = (componentName, config, configTypes) => {
...
 
 const isVisible = element => {
-  if (!element) {
+  const eligibleElement = el => el && el.style
+
+  if (!eligibleElement(element)) {
+    return false
+  }
+
+  if (element.tagName !== 'AREA' && element.getClientRects().length === 0) {
```
Comment (GeoSot) [inline js/src/util/index.js:144]: I think SVG too, returns empty `getClientRects()`

File: js/src/util/index.js [hunk 4]
```diff
@@ -135,20 +135,33 @@ const typeCheckConfig = (componentName, config, configTypes) => {
...
-  if (element.style && element.parentNode && element.parentNode.style) {
-    const elementStyle = getComputedStyle(element)
-    const parentNodeStyle = getComputedStyle(element.parentNode)
+  let relevantVisibility
+
+  while (eligibleElement(element)) {
+    const { display, visibility } = getComputedStyle(element)
+
+    if (!relevantVisibility && /visible|hidden/.test(visibility)) {
+      relevantVisibility = visibility
```
Comment (GeoSot) [inline js/src/util/index.js:154]: A bit tricky check IMO .  I saw the extra test and it is valid, however I think it is maybe an edge case, of  bad use

File: js/src/util/index.js (hunk 3 above)
Comment (alpadev) [reply js/src/util/index.js:144]: Well at least not in Chrome.. even a `<clipPath>` is returning one.

File: js/src/util/index.js (hunk 4 above)
Comment (alpadev) [reply js/src/util/index.js:154]: According to the CSS spec:

> The visibility property specifies whether the boxes generated by an element are rendered. Invisible boxes still affect layout (set the display property to none to suppress box generation altogether). Values have the following meanings:
>
> - visible
> The generated box is visible.
> - hidden
> The generated box is invisible (fully transparent, nothing is drawn), but still affects layout. Furthermore, descendants of the element will be visible if they have 'visibility: visible'.

File: js/src/util/index.js (hunk 3 above)
Comment (GeoSot) [reply js/src/util/index.js:144]: after a second thought, mayby we overthink it. It is for internal use, and by default checks buttons and component elements. So maybe we can stick on basics. Nor svg/area/clippath etc .

it we find an edge case we can support it later. Either way, the check is better than before :smile:

**Comment by RyanBerliner** [comment]
> @RyanBerliner Hm wouldn't be checking `getClientRects()` enough to see if an element (or an ancestor of it) has `display: none`? 🤔

99% of the time, yes! However, so long as we keep the `element.tagName !== 'AREA'`, then we'll still need to check the display property for supporting that specific tag name. I know we could also squeeze in a few more edge cases to support here, though I'd imagine it's preferred only to improve it as much as we think the framework will need it. I'm looking at this function through the lens of most rigorous use case bootstrap will be using it for, which is probably #33865.

This is why I've decided to support the `<area>` edge case, and not (for example) the svg edge case. `<area>` is focusable by default, and I'd think someone is more likely to put an image map in a modal (still unlikely though lol) than an svg with an explicit tabindex. If you all have more insight into what edge cases you'd imagine should be supported I can adjust accordingly.

File: js/src/util/index.js (hunk 1 above)
Comment (XhmikosR) [reply js/src/util/index.js:140]: It's subjective, I'd say, but to me it's the same plus it's shorter :)

File: js/src/util/index.js (hunk 3 above)
Comment (GeoSot) [inline js/src/util/index.js:144]: Please add a comment explaining the 'AREA' check for future reference

**Review by GeoSot — COMMENTED** [review]
Maybe `window.getComputedStyle(x).getPropertyValue('visibility')` would work without tricks 
~~[Example](https://codepen.io/GeosSV/pen/abJmGGw) wrong link sorry~~
[Example](https://codepen.io/GeosSV/pen/abJmGGw)

`  const eligibleElement = el => el && el.style` could by replaced be `isElement`?

**Comment by alpadev** [comment]
@RyanBerliner From my observations `getClientRects()` seem to work for `<area>`. That's true at least with the latest versions of Chromium, FF and Safari. I did some research and found a Chromium bug, that mentions, that they eventually changed that some years ago to match the FF behaviour. (https://bugs.chromium.org/p/chromium/issues/detail?id=416195#c6)

Also I found that even when you apply `display: none` to an `<area>` or `<map>` element, you can still tab to it and click it. So not sure if it makes sense even, to check if they are visible or not.

Not sure if we should just risk it, and fix it if some issue pops up.

**Comment by RyanBerliner** [comment]
Interesting - I'm starting to think we should just completely remove the edge case handling of `<area>` as it seems it's a little all over the place with browser behavior. [Here is where I got my initial inspiration to handle `<area>` separately, fyi.](https://developer.mozilla.org/en-US/docs/Web/API/Element/getClientRects) Then we can completely remove the loop, rely entirely on `getClientRect()`, and do a final visibility check with the computed style property like @GeoSot suggested. Would that work with everyone or should I proceed with the other suggested changes while maintaining separate `<area>` checking?

**Comment by alpadev** [comment]
To be honest, I don't really have a preference here. It's hard to argue when neither the browser nor the specification is clear about this behavior. I feel like image maps may be some abandoned thing since they lack of responsiveness or something.

I had a talk with Geo yesterday, and he asked about opacity. My guess that would require to have a loop, if we want to take care about that too.

Hard to say what's the best approach here, but I'm a bit concerned, that with that many iterations we will block you or kill your mood a bit. :heart:

**Comment by GeoSot** [comment]
IMO kill the 'AREA' thing. Is so rare.
The checks are already better than before. 


As @alpadev said:
 >Maybe we should just risk it, and fix it if some issue pops up.
 
 I cannot agree more.

File: js/tests/unit/util/index.spec.js [hunk 5]
```diff
@@ -324,22 +324,6 @@ describe('Util', () => {
       expect(Util.isVisible(div)).toEqual(false)
     })
 
-    it('should return true if an ancestor element is visibility hidden, but reverted', () => {
```
Comment (GeoSot) [inline js/tests/unit/util/index.spec.js]: Based on this https://codepen.io/GeosSV/pen/abJmGGw, your test should work.

File: js/src/util/index.js [hunk 6]
```diff
@@ -135,33 +135,15 @@ const typeCheckConfig = (componentName, config, configTypes) => {
...
 const isVisible = element => {
-  const eligibleElement = el => el && el.style
-
-  if (!eligibleElement(element)) {
-    return false
-  }
-
-  if (element.tagName !== 'AREA' && element.getClientRects().length === 0) {
+  if (
+    !element ||
```
Comment (GeoSot) [inline js/src/util/index.js:139]:
```suggestion

```

I think `isElement()` check is enough

File: js/tests/unit/util/index.spec.js (hunk 5 above)
Comment (RyanBerliner) [reply js/tests/unit/util/index.spec.js]: Correct, that test would have passed based on how the computed visibility works. Would you like me to add back this test?

File: js/src/util/index.js (hunk 6 above)
Comment (RyanBerliner) [reply js/src/util/index.js:139]: Unfortunately `isElement()` expects a non-null parameter, and there is an existing `isVisible()` test requiring safe handling of a `null|undefined` parameter. While adding support for a null parameter to `isElement()` seems like the right thing to do, I'd imagine it should be done outside this PR for clarity purposes.

File: js/tests/unit/util/index.spec.js (hunk 5 above)
Comment (GeoSot) [reply js/tests/unit/util/index.spec.js]: why not. it was nice :smiley:

File: js/src/util/index.js (hunk 6 above)
Comment (GeoSot) [reply js/src/util/index.js:139]: Are you sure? :thinking: 

https://github.com/twbs/bootstrap/blob/de0b797ab2bb80e4576c15f1992dc5755378edcd/js/src/util/index.js#L106
Looks to work


image

Comment (RyanBerliner) [reply js/src/util/index.js:139]: oh lol - this branch has not been synced since you fixed it in #33327, thanks! I'll sync and update 👍

**Review by GeoSot — APPROVED** [review]
